apps/main-site/src/pages/api/sitemap-checker.ts

```diff
@@ -0,0 +1,58 @@
+/* eslint-disable @typescript-eslint/restrict-template-expressions */
+/* eslint-disable @typescript-eslint/no-unsafe-argument */
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+import { NextApiRequest, NextApiResponse } from 'next';
+import { parseStringPromise } from 'xml2js';
+export default function handler(req: NextApiRequest, res: NextApiResponse) {
+	void fetchSitemap();
+}
+
+async function fetchSitemap() {
+	// fetching answeroverflow sitemap
+	const startTime = new Date();
+	const sitemapResponse = await fetch('https://Answeroverflow.com/sitemap.xml');
+	const endTime = new Date();
+	console.log(
+		`Time taken to fetch https://Answeroverflow.com/sitemap.xml: ${
+			endTime.getTime() - startTime.getTime()
+		}ms`,
+	);
+
+	// parsing answeroverflow sitemap's xml to js
+	const sitemapText = await sitemapResponse.text();
+	const data = await parseStringPromise(sitemapText);
+
+	// Fetching and parsing each sitemap then retrieving all urls from each sitemap
+	let sitemapCount = 0;
+	let urlCount = 0;
+	const startTime2 = new Date();
+	console.log('Fetching and parsing each sitemap...');
+	const promises = data.sitemapindex.sitemap.map(async (sitemap) => {
+		const resp = await fetch(sitemap.loc[0]);
+		const textData = await resp.text();
+		const xmlDoc = await parseStringPromise(textData);
+
+		await Promise.all(
```

Comment:
Why are we using a promise here?

Reply:
using promises because raw await in loops is not allowed by eslint, but with this i do not need to disable "no-await-in-loop"

Reply:
upto my knowledge, 
first i didn't used for loop because we do not know the condition when it will end 
second on using (for of) await was not allowed
And since await is not allowed in loop so i searched what can i do then i found the commited solution to it (using map and promise.all).

and using async function so that fetch can be awaited inside the loop because we are "fetching" each sitemap which is necessary to obtain all the urls/pages it has (after fetching we need to await to parse the fetched data also, so we need await there.)

but i don't know lot of stuff, this was all i was able to figure out on my own.

Reply:
@JoltCode

Reply:
I'm talking about your loop inside the loop, that's only incrementing a number, not the outer loop.

Reply:
@JoltCode 
ohh sorry i overlooked telling you something,
We are using promise.all two times
the reason is since we are **fetching large amount of urls and sitemaps**, something which can be done asynchronsly for improving the performance.
Earlier it was taking me let say 40 seconds,
but now with promise.all which **enablles parallel execution** made our task async took like 5 second.

Promise.all just takes array or map, to perform the async task, and thats why map.